src/tests.rs

```diff
@@ -627,40 +632,84 @@ async fn test_key_extraction_whitelisted_key_with_header() {
             .route("/", web::get().to(hello)),
     )
     .await;
-    let create_req_and_test = |user_token: HeaderValue| async {
+    let create_res = |user_token: HeaderValue| async {
         let mut req = test::TestRequest::get().uri("/").to_request();
         req.headers_mut()
             .insert(HeaderName::from_static("super-token"), user_token);
         let test = test::call_service(&app, req).await;
-        dbg!(test.headers());
-        assert_eq!(test.status(), StatusCode::OK);
-        assert_eq!(
-            test.headers()
-                .get(HeaderName::from_static("x-ratelimit-whitelisted"))
-                .unwrap(),
-            "true"
-        );
-        assert!(test
-            .headers()
-            .get(HeaderName::from_static("x-ratelimit-limit"))
-            .is_none());
-        assert!(test
-            .headers()
-            .get(HeaderName::from_static("x-ratelimit-remaining"))
-            .is_none());
-        assert!(test
-            .headers()
-            .get(HeaderName::from_static("x-ratelimit-after"))
-            .is_none());
+        (test.headers().clone(), test.status().as_u16())
     };
-    // First request
-    create_req_and_test(HeaderValue::from_static("AdminSecretToken")).await;
-    // Second request
-    create_req_and_test(HeaderValue::from_static("AdminSecretToken")).await;
-    // Third request
-    create_req_and_test(HeaderValue::from_static("AdminSecretToken2")).await;
-    // Fourth request
-    create_req_and_test(HeaderValue::from_static("AdminSecretToken2")).await;
+    #[rustfmt::skip]
+    let test_res = move |(headers, status): (HeaderMap, u16)| {
+        move |
+            whitelisted: Option<&HeaderValue>,
+            limit: Option<&HeaderValue>,
+            remaining: Option<&HeaderValue>,
+            after: Option<&HeaderValue>,
+            excepted_status: u16| {
+            assert_eq!(status, excepted_status);
+            assert_eq!(headers.get("x-ratelimit-whitelisted"), whitelisted);
+            assert_eq!(headers.get("x-ratelimit-limit"), limit);
+            assert_eq!(headers.get("x-ratelimit-remaining"), remaining);
+            assert_eq!(headers.get("x-ratelimit-after"), after);
+        }
+    };
```

Comment:
Can't you just summarize this into once function? I don't understand why you need two closures here.

Also, FYI [happy path](https://en.wikipedia.org/wiki/Happy_path) is an actual term used in CS :)

Reply:
> Also, FYI [happy path](https://en.wikipedia.org/wiki/Happy_path) is an actual term used in CS :)

Ohh 😆. New to me